platform/src/main/java/org/hillview/sketches/Centroids.java

```diff
@@ -0,0 +1,105 @@
+     */
+    public HashMap<T, double[]> sums;
+    /**
+     * Map from the partition key to the count of values in every column of that partition.
+     */
+    public HashMap<T, long[]> counts;
+
+    /**
+     * Constructs a zero-centroids.
+     */
+    public Centroids() {
+        this.sums = new HashMap<T, double[]>();
+        this.counts = new HashMap<T, long[]>();
+    }
+
+    /**
+     * Construct a centroids object that has processed the sums of the partitions in 'table'.
+     * The final centroids are not computed yet, this should be done after the sketch finishes.
+     * @param members MembershipSet that has knows which rows to iterate over.
+     * @param keyFunc Function that can determine the partition key of a row entry.
+     * @param columns Column that define the nD space.
+     */
+    public Centroids(IMembershipSet members, Function<Integer, T> keyFunc, List<IColumn> columns) {
+        this();
+        int colIndex = 0;
+        for (IColumn column : columns) {
+            IRowIterator rowIterator = members.getIterator();
+            int row = rowIterator.getNextRow();
+            while (row >= 0) {
+                if (column.isMissing(row))
+                    continue;
+                T key = keyFunc.apply(row);
+                // Update the values. (And first create them if absent.)
+                this.sums.computeIfAbsent(key, k -> new double[columns.size()])[colIndex] += column.getDouble(row);
+                this.counts.computeIfAbsent(key, k -> new long[columns.size()])[colIndex]++;
+                row = rowIterator.getNextRow();
+            }
+            colIndex++;
+        }
+    }
+
+    /**
+     * Compute the aggregate of two centroid objects. This simply sums the sums and counts.
+     * @param other Object that holds information for the other set of centroids.
+     * @return Centroid object that represents the aggregate of the rows that both objects represent.
+     */
+    public Centroids<T> union(Centroids<T> other) {
+        other.sums.keySet().forEach((key) -> {
+            if (this.sums.containsKey(key)) {
+                double[] sum = other.sums.get(key);
+                // other.counts must also contain the key (they share key sets, no need to check or loop twice).
+                long[] count = other.counts.get(key);
+                for (int colIndex = 0; colIndex < sum.length; colIndex++) {
+                    this.sums.get(key)[colIndex] += sum[colIndex];
+                    // this.count must also contain the key as well, as it shares key set with this.sum.
+                    this.counts.get(key)[colIndex] += count[colIndex];
+                }
+            } else {
+                // No need to iterate/sum. Simply use the other's array.
+                this.sums.put(key, other.sums.get(key));
+                this.counts.put(key, other.counts.get(key));
+            }
+        });
+        return this;
```

Comment:
Please don't mutate this; we want everything to be immutable and thread-safe; allocate a new Centroids.
Also, it looks like you are losing elements in this which are not in other.